File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def condition(self,piles: List[int],k: int ,h: int) -> bool:
        score = 0
        for pile in piles:
            score += math.ceil(pile / k)
        
        return score <= h

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        """
            eating rate is k

            all bananas within h hours

            [1,4,3,2]
            h = 9

            k = 2
            h = 0

            for pile in piles:
                h += ceil(pile / k)

            h = 1 [1,4,3,2]
            1 - 1
            4 - 2
            3 - 2
            2 - 

            while left <= right:
                mid = left + (right-left)//2

                if condition(piles,mid,h):
                    right = mid
                
                else:
                    left = mid + 1

            return right

            [25,10,23,4]

        """
        left = 1
        right = max(piles)

        while left < right:
            mid = left + (right-left)//2

            if self.condition(piles,mid,h):
                right = mid
                
            else:
                left = mid + 1

        return right

File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int n = heights.size();
        int left = 0;
        int right = n-1;
        /*
            1   7   2   5   4   7   3   6
        l       |
        r                               |

            if 1 < 6 then move right
            if 7 > 6 then 
        */
        int area = 0;
        while(left < right){
            int currArea = min(heights[left],heights[right]) * (right - left);
            area = max(area,currArea);
            if(heights[left] < heights[right]){
                left++;
            }
            else{
                right--;
            }

        }
        return area;
    }
};
